zerocopy-derive/src/enums.rs

```diff
@@ -0,0 +1,362 @@
+                    //   constant for this variant, so this cast preserves
+                    //   types and locations of all fields. Therefore, any
+                    //   `UnsafeCell`s will have the same location as in the
+                    //   original type.
+                    let variant = unsafe {
+                        variants.cast_unsized(
+                            |p: *mut ___ZerocopyVariants #ty_generics| {
+                                p as *mut #variant_struct_ident #ty_generics
+                            }
+                        )
+                    };
+                    // SAFETY: `cast_unsized` removes the initialization
+                    // invariant from `p`, so we re-assert that all of the bytes
+                    // are initialized.
+                    let variant = unsafe { variant.assume_initialized() };
+                    <
+                        #variant_struct_ident #ty_generics as #trait_path
+                    >::is_bit_valid(variant)
+                }
+            }
+        }
+    });
+
+    quote! {
+        // SAFETY: We use `is_bit_valid` to validate that the bit pattern of the
+        // enum's tag corresponds to one of the enum's discriminants. Then, we
+        // check the bit validity of each field of the corresponding variant.
+        // Thus, this is a sound implementation of `is_bit_valid`.
+        fn is_bit_valid<A>(
+            mut candidate: ::zerocopy::Maybe<'_, Self, A>,
+        ) -> ::zerocopy::macro_util::core_reexport::primitive::bool
+        where
+            A: ::zerocopy::pointer::invariant::Aliasing
+                + ::zerocopy::pointer::invariant::AtLeast<::zerocopy::pointer::invariant::Shared>,
+        {
+            use ::zerocopy::macro_util::core_reexport;
+
+            #tag_enum
+
+            type ___ZerocopyTagPrimitive = ::zerocopy::macro_util::SizeToTag<{
+                core_reexport::mem::size_of::<___ZerocopyTag>()
+            }>;
+
+            #tag_consts
+
+            type ___ZerocopyOuterTag = #outer_tag_type;
+            type ___ZerocopyInnerTag = #inner_tag_type;
+
+            #variant_structs
+
+            #variants_union
+
+            #[repr(C)]
+            struct ___ZerocopyRawEnum #generics {
+                tag: ___ZerocopyOuterTag,
+                variants: ___ZerocopyVariants #ty_generics,
+            }
+
+            let tag = {
+                // SAFETY:
+                // - The provided cast addresses a subset of the bytes addressed
+                //   by `candidate` because it addresses the starting tag of the
+                //   enum. Because the pointer is cast from `candidate`, it has
+                //   the same provenance as it. There are no `UnsafeCell`s in
+                //   the tag because it is a primitive integer.
+                // - By definition, `Self` and `__ZerocopyTagPrimitive` are
+                //   types of the same size.
```

Comment:
```suggestion
                // - The size of `__ZerocopyTagPrimitive` is less-than-or-equal
                // to that of `Self`.
```
`__ZerocopyTagPrimitive` will be smaller if `Self` is a dataful enum, right?

Reply:
I think that bullet point is vestigial, I reworded the first bullet point to cover all of the points and didn't delete the second. Removing that in the next push.